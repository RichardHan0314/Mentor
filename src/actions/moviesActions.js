import axios from "axios";

const ROOT_URL = `http://localhost:4000`;

export const GET_MOVIES = 'GET_MOVIES'; // положили строку в константу
export const GET_MOVIES_REQUEST = 'GET_MOVIES_REQUEST';
export const GET_MOVIES_SUCCESS = 'GET_MOVIES_SUCCESS';

const getMovies = data => (dispatch, getState) => {
    const url = `${ROOT_URL}/movies`;
    const request = axios.get(url);

    request.then(response => {
        dispatch({
            type: GET_MOVIES_REQUEST,
            payload: response.data.data
        });
    });
};

export default getMovies;

