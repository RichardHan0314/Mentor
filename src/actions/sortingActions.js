import { getAllMovies, sortValueSelector } from '../selectors';

export const SET_SORTING_VALUE = 'SET_SORTING_VALUE';
export const SORT_MOVIES = 'SORT_MOVIES';

export const setSortingValue = value => ({
    type: SET_SORTING_VALUE,
    payload: value,
});

export const sortMovies = () => (dispatch, getState) => {
    let movies = getAllMovies(getState());
    const sortBy = sortValueSelector(getState());

    if (sortBy === 'release_date') {
        movies = [...movies].sort((a, b) => {
            if (new Date(b[sortBy]) < new Date(a[sortBy])) return -1;
            if (new Date(b[sortBy]) > new Date(a[sortBy])) return 1;
            return 0;
        });
    } else {
        movies = [...movies].sort((a, b) => b[sortBy] - a[sortBy]);
    }

    return dispatch({
        type: SORT_MOVIES,
        payload: movies,
    });
};
