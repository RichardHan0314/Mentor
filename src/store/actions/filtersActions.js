import { getAllFilters } from '../selectors';

export const GET_FILTERS = 'GET_FILTERS';
export const SET_ALL_FILTERS = 'SET_ALL_FILTERS';
export const SET_FILTERS = 'SET_FILTERS';

export const getFilters = (getState) => {
    const allFilters = getAllFilters(getState());
    return {
        type: GET_FILTERS,
        payload: allFilters,
    };
};

export const setAllFilters = () => (dispatch, getState) => {
    const allFilters = getAllFilters(getState());
    const newFilters = [...allFilters];
    newFilters.map(filter => ({ ...filter, status: false }));

    dispatch({
        type: SET_ALL_FILTERS,
        payload: { allFilters: newFilters, allFiltersChecked: true},
    });
};

export const setFilters = filterIndex => (dispatch, getState) => {
    const allFilters = getAllFilters(getState());
    allFilters[filterIndex].status = !allFilters[filterIndex].status;
    const newFilters = [...allFilters];

    dispatch({
        type: SET_FILTERS,
        payload: { allFilters: newFilters, allFiltersChecked: false},
    });
};
