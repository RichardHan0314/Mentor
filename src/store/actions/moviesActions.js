import axios from "axios";
import { searchValueSelector, getFilters, sortValueSelector, getAllMovies } from '../selectors';

const API_GATEWAY = `http://localhost:4000/movies`;
const SUCCESS_DELETE_CODE = 204;
const SUCCESS_ADD_CODE = 201;
const SUCCESS_UPDATE_CODE = 200;

export const GET_MOVIES = 'GET_MOVIES';
export const GET_MOVIES_REQUEST = 'GET_MOVIES_REQUEST';
export const GET_MOVIES_REQUEST_START = 'GET_MOVIES_REQUEST_START';
export const GET_MOVIES_REQUEST_SUCCESS = 'GET_MOVIES_REQUEST_SUCCESS';
export const GET_MOVIES_REQUEST_ERROR = 'GET_MOVIES_REQUEST_ERROR';
export const ADD_MOVIE = 'ADD_MOVIE';
export const UPDATE_MOVIE = 'UPDATE_MOVIE';
export const DELETE_MOVIE = 'DELETE_MOVIE';

const getMovies = () => (dispatch, getState) => {
    const params = {};
    params.search = searchValueSelector(getState());
    params.searchBy = 'title';
    params.sortBy = sortValueSelector(getState());
    params.filter = getFilters(getState());
    console.log('params', params);

    const request = axios.get(API_GATEWAY, {params});

    request.then(response => {
        dispatch({
            type: GET_MOVIES_REQUEST,
            payload: response.data.data
        });
    });
};

export const requestMovies = options => async (dispatch) => {
    dispatch({
        type: GET_MOVIES_REQUEST_START,
        payload: { loading: true }
    })
    try {
        const { data } = await axios.get(API_GATEWAY, options);

        dispatch({
            type: GET_MOVIES_REQUEST_SUCCESS,
            payload: { movies: data.data, loading: false },
        })
    } catch (error) {
        dispatch({
            type: GET_MOVIES_REQUEST_ERROR,
            payload: { error, loading: false },
        })
    }
}

export const addMovie = (movieData) => (dispatch, getState) => {
    const movies = getAllMovies(getState());
    const request = axios.post(API_GATEWAY, {...movieData});

    request.then(response => {
        if (response.status === SUCCESS_ADD_CODE) {
            dispatch({
                type: ADD_MOVIE,
                payload: [response.data, ...movies]
            });
        };
    });
};

export const updateMovie = (movieData) => (dispatch, getState) => {
    const movies = getAllMovies(getState());
    const request = axios.put(API_GATEWAY, {...movieData});

    request.then(response => {
        if (response.status === SUCCESS_UPDATE_CODE) {
            const  index = movies.findIndex(x => x.id === response.data.id);
            movies.splice(index, 1, response.data);
            const newMovies = [...movies];

            dispatch({
                type: UPDATE_MOVIE,
                payload: newMovies
            });
        };
    });
};

export const deleteMovies = (movieId) => (dispatch, getState) => {
    const url = `${API_GATEWAY}/${movieId}`;
    const movies = getAllMovies(getState());
    const newMovies = [...movies].filter(item => item.id !== movieId);

    const request = axios.delete(url);

    request.then(response => {
        if (response.status === SUCCESS_DELETE_CODE) {
            dispatch({
                type: DELETE_MOVIE,
                payload: newMovies
            });
        }
    });
};

export default getMovies;
