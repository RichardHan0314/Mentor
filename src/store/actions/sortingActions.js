export const SET_SORTING_VALUE = 'SET_SORTING_VALUE';
export const SORT_MOVIES = 'SORT_MOVIES';

export const setSortingValue = value => ({
    type: SET_SORTING_VALUE,
    payload: value,
});

export const sortMovies = (sortedMovies) => ({
    type: SORT_MOVIES,
    payload: sortedMovies,
});
