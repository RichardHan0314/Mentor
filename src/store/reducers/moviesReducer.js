/* eslint-disable import/prefer-default-export */
import { ADD_MOVIE, UPDATE_MOVIE, DELETE_MOVIE_REQUEST_SUCCESS, GET_MOVIES_REQUEST_SUCCESS } from '../actions/moviesActions';
import { SORT_MOVIES } from '../actions/sortingActions';

const initialState = {
    movies: []
};

export function moviesReducer(state = initialState, action) {
    switch (action.type) {
    case GET_MOVIES_REQUEST_SUCCESS:
        return {...state, movies: action.payload.movies };
    case SORT_MOVIES:
        return {...state, movies: action.payload };
    case ADD_MOVIE:
        return {...state, movies: action.payload };
    case UPDATE_MOVIE:
        return {...state, movies: action.payload };
    case DELETE_MOVIE_REQUEST_SUCCESS:
        return {
            ...state,
            movies: [...state.movies].filter(movie => movie.id !== action.payload.movieId)
        };
    default:
        return state;
    }
}
