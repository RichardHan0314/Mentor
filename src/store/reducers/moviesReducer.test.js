import { moviesReducer } from './moviesReducer';
import { 
    GET_MOVIES_REQUEST_SUCCESS,
    ADD_MOVIE_REQUEST_SUCCESS,
    UPDATE_MOVIE_REQUEST_SUCCESS,
    DELETE_MOVIE_REQUEST_SUCCESS
} from '../actions/moviesActions';
import { SORT_MOVIES } from '../actions/sortingActions';

const moviesMock = [{
    id: 1,
    title: 'Aaaa'
},
{
    id: 2,
    title: 'Bbbb'
}];

describe('filters_reducer', () => {
    describe('INITIAL_STATE', () => {
        test('is correct', () => {
            const action = { type: 'dummy_action' };
            expect(moviesReducer(undefined, action)).toMatchSnapshot();
        });
    });
      
    describe('GET_MOVIES_REQUEST_SUCCESS', () => {
        test('returns the correct state', () => {
            const action = { type: GET_MOVIES_REQUEST_SUCCESS, payload: moviesMock };

            expect(moviesReducer(undefined, action)).toMatchSnapshot();
        });
    });

    describe('ADD_MOVIE_REQUEST_SUCCESS', () => {
        test('returns the correct state', () => {
            const action = { type: ADD_MOVIE_REQUEST_SUCCESS, payload: moviesMock };

            expect(moviesReducer(undefined, action)).toMatchSnapshot();
        });
    });

    describe('UPDATE_MOVIE_REQUEST_SUCCESS', () => {
        test('returns the correct state', () => {
            const action = { type: UPDATE_MOVIE_REQUEST_SUCCESS, payload: moviesMock };

            expect(moviesReducer(undefined, action)).toMatchSnapshot();
        });
    });

    describe('DELETE_MOVIE_REQUEST_SUCCESS', () => {
        test('returns the correct state', () => {
            const movieIdMock = 1;
            const action = { type: DELETE_MOVIE_REQUEST_SUCCESS, payload: movieIdMock };

            const initialState = {
                movies: moviesMock
            };

            expect(moviesReducer(undefined, action)).toMatchSnapshot();

            const movies = [...initialState.movies].filter(movie => movie.id !== action.payload.movieIdMock);

            expect(moviesReducer(initialState, action)).toEqual({
                ...initialState,
                movies,
            });
        });
    });

    describe('SORT_MOVIES', () => {
        test('returns the correct state', () => {
            const action = { type: SORT_MOVIES, payload: moviesMock };

            expect(moviesReducer(undefined, action)).toMatchSnapshot();
        });
    });
});
