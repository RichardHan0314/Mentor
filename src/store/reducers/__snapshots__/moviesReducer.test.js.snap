// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filters_reducer ADD_MOVIE_REQUEST_SUCCESS returns the correct state 1`] = `
Object {
  "movies": Array [
    undefined,
  ],
}
`;

exports[`filters_reducer DELETE_MOVIE_REQUEST_SUCCESS returns the correct state 1`] = `
Object {
  "movies": Array [],
}
`;

exports[`filters_reducer GET_MOVIES_REQUEST_SUCCESS returns the correct state 1`] = `
Object {
  "movies": undefined,
}
`;

exports[`filters_reducer INITIAL_STATE is correct 1`] = `
Object {
  "movies": Array [],
}
`;

exports[`filters_reducer SORT_MOVIES returns the correct state 1`] = `
Object {
  "movies": Array [
    Object {
      "id": 1,
      "title": "Aaaa",
    },
    Object {
      "id": 2,
      "title": "Bbbb",
    },
  ],
}
`;

exports[`filters_reducer UPDATE_MOVIE_REQUEST_SUCCESS returns the correct state 1`] = `
Object {
  "movies": undefined,
}
`;
